### PR TITLE
Refine style guide

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of docs revamp #7666

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7

### CODE REVIEW

File: docs/source/starter/style_guide.rst [hunk 1]
```diff
@@ -24,7 +26,7 @@ In Lightning we differentiate between a system and a model.
 
 A model is something like a resnet18, RNN, etc.
 
-A system defines how a collection of models interact with each other. Examples of this are:
+A system defines how a collection of models interact with each other with user-defined training/evaluation logic. Examples of this are:
 
 * GANs
 * Seq2Seq
```
Comment (akihironitta) [inline docs/source/starter/style_guide.rst:32]: The below bullet could be simply deleted.
```
* etc
```

File: docs/source/starter/style_guide.rst [hunk 2]
```diff
@@ -198,7 +211,20 @@ because all these questions need to be answered:
...
+------------
+
+********
+Examples
+********
+
+Checkout the live examples to get your hands dirty:
+
+- `Introduction to PyTorch Lightning <https://pytorch-lightning.readthedocs.io/en/latest/notebooks/lightning_examples/mnist-hello-world.html>`_
+- `Introduction to DataModules <https://pytorch-lightning.readthedocs.io/en/latest/notebooks/lightning_examples/datamodules.html>`_
```
Comment (akihironitta) [inline docs/source/starter/style_guide.rst:230]: Following Jirka's comment in https://github.com/PyTorchLightning/pytorch-lightning/pull/10440#discussion_r773263578, let's use `stable` instead of `latest`.
```suggestion
- `Introduction to PyTorch Lightning <https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/mnist-hello-world.html>`_
- `Introduction to DataModules <https://pytorch-lightning.readthedocs.io/en/stable/notebooks/lightning_examples/datamodules.html>`_
```

File: docs/source/starter/style_guide.rst (hunk 1 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:32]: why ? 🤔

File: docs/source/starter/style_guide.rst [hunk 3]
```diff
@@ -49,7 +51,7 @@ Here's a LightningModule that defines a system:
 .. testcode::
 
     class LitModel(LightningModule):
-        def __init__(self, encoder: nn.Module = None, decoder: nn.Module = None):
```
Comment (ananthsub) [inline docs/source/starter/style_guide.rst:52]: n00b question: why remove the typehint?

File: docs/source/starter/style_guide.rst [hunk 4]
```diff
@@ -147,26 +153,27 @@ In practice, this code looks like:
...
 
 .. code-block:: python
 
-    def training_step(self):
-        x, y = ...
+    def training_step(self, batch, batch_idx):
+        x, y = batch
         z = self(x)  # < ---------- instead of self.encoder(x)
         pred = self.decoder(z)
         ...
```
Comment (ananthsub) [inline docs/source/starter/style_guide.rst:179]: should we recommend users to implement forward then to avoid a special carveout for DataParallel? it's not intuitive why DataParallel requires this

File: docs/source/starter/style_guide.rst [hunk 5]
```diff
@@ -176,19 +183,25 @@ However, when using DataParallel, you will need to call forward manually
...
 
-.. warning:: Make sure not to use ddp_spawn with num_workers > 0 or you will bottleneck your code.
+.. warning:: Make sure not to use ``Trainer(strategy="ddp_spawn")`` with ``num_workers>0`` in a DataLoader or you will bottleneck your code.
 
 DataModules
 ===========
-Lightning introduced datamodules. The problem with dataloaders is that sharing full datasets is often still challenging
-because all these questions need to be answered:
+
+Lightning introduced :class:`~pytorch_lightning.core.datamodule.LightningDataModule` in v0.9 release. They are designed as a way of decoupling data-related
```
Comment (ananthsub) [inline docs/source/starter/style_guide.rst:200]: the version it was introduced in is not so useful imo

File: docs/source/starter/style_guide.rst [hunk 6]
```diff
@@ -1,7 +1,8 @@
 ###########
-Style guide
+Style Guide
 ###########
-A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo,
+
+A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo or a research project,
 finding a lightning module and knowing exactly where to look to find the things you care about.
```
Comment (ananthsub) [inline docs/source/starter/style_guide.rst:7]:
```suggestion
A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo or a research project,
finding a lightning module, and knowing exactly where to look to find the things you care about.

```

File: docs/source/starter/style_guide.rst (hunk 1 above)
Comment (akihironitta) [reply docs/source/starter/style_guide.rst:32]: It just looks unnecessary to me, but if you think it looks nicer as is, let's keep it :)

File: docs/source/starter/style_guide.rst [hunk 7]
```diff
@@ -38,7 +40,7 @@ Here's a LightningModule that defines a model:
 .. testcode::
 
     class LitModel(LightningModule):
-        def __init__(self, num_layers: int = 3):
+        def __init__(self):
```
Comment (ananthsub) [inline docs/source/starter/style_guide.rst:43]: "A LightningModule can define both a system and a model" defeats the point above. 
Users should define pure nn.Modules *outside* of the LightningModule. Including the examples with the layers like this isn't how I've seen the vast majority of users interact with Lightning

File: docs/source/starter/style_guide.rst (hunk 5 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:200]: will update!

File: docs/source/starter/style_guide.rst (hunk 7 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:43]: will update!

File: docs/source/starter/style_guide.rst (hunk 3 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:52]: we don't usually have typehints in doc examples.

File: docs/source/starter/style_guide.rst (hunk 7 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:43]: It's a choice I'd say... although a system is what we recommend but if it's a very simple model, then we don't enforce it.

File: docs/source/starter/style_guide.rst (hunk 4 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:179]: DP is a special case worth highlighting. Advising implementing a forward method always might not be good for this section.

File: docs/source/starter/style_guide.rst (hunk 3 above)
Comment (awaelchli) [reply docs/source/starter/style_guide.rst:52]: yes agreed

File: docs/source/starter/style_guide.rst [hunk 8]
```diff
@@ -176,19 +189,25 @@ However, when using DataParallel, you will need to call forward manually
...
 DataModules
 ===========
-Lightning introduced datamodules. The problem with dataloaders is that sharing full datasets is often still challenging
-because all these questions need to be answered:
+
+The :class:`~pytorch_lightning.core.datamodule.LightningDataModule` is designed as a way of decoupling data-related
+hooks from the :class:`~pytorch_lightning.core.lightning.LightningModule` so you can develop dataset agnostic models. It makes it easy to hot swap different
+datasets with your model, so you can test it and benchmark it across domains. It also makes sharing and reusing the exact data splits and transforms across projects possible.
+
+The problem with dataloaders is that sharing full datasets is often still challenging because all these questions need to be answered:
```
Comment (awaelchli) [inline docs/source/starter/style_guide.rst:210]: This does not longer fit well since you have introduced the concept already above. So the impact of asking these questions is lost. Let's make sure the flow of the intro does not break here.

File: docs/source/starter/style_guide.rst [hunk 9]
```diff
@@ -198,7 +217,20 @@ because all these questions need to be answered:
 It's for this reason that we recommend you use datamodules. This is specially important when collaborating because
 it will save your team a lot of time as well.
 
-All they need to do is drop a datamodule into a lightning trainer and not worry about what was done to the data.
+All they need to do is drop a datamodule into a :class:`~pytorch_lightning.trainer.trainer.Trainer` and not worry about what was done to the data.
 
 This is true for both academic and corporate settings where data cleaning and ad-hoc instructions slow down the progress
 of iterating through ideas.
+
+Checkout :ref:`data` doc to understand data management within Lightning and it's best practices.
```
Comment (awaelchli) [inline docs/source/starter/style_guide.rst:225]:
```suggestion
Checkout :ref:`data` doc to understand data management within Lightning and its best practices.
```

File: docs/source/starter/style_guide.rst (hunk 8 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst]: removed the long intro.. I guess the reference is enough.

File: docs/source/starter/style_guide.rst [hunk 10]
```diff
@@ -176,29 +189,33 @@ However, when using DataParallel, you will need to call forward manually
...
 
-* What splits were used?
-* How many samples does this dataset have?
-* What transforms were used?
-* etc...
+The :class:`~pytorch_lightning.core.datamodule.LightningDataModule` is designed as a way of decoupling data-related
+hooks from the :class:`~pytorch_lightning.core.lightning.LightningModule` so you can develop dataset agnostic models. It makes it easy to hot swap different
+datasets with your model, so you can test it and benchmark it across domains. It also makes sharing and reusing the exact data splits and transforms across projects possible.
+
+Checkout :ref:`data` document to understand data management within Lightning and its best practices.
```
Comment (awaelchli) [inline docs/source/starter/style_guide.rst:210]:
```suggestion
Check out :ref:`data` document to understand data management within Lightning and its best practices.
```
Quiz:
Which of the following is a valid git command?

A) git check out
B) git checkout
C) git chickenout

**Review by awaelchli — APPROVED** [review]

File: docs/source/starter/style_guide.rst (hunk 10 above)
Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:210]: C) :smile: 

oh! man. I swear I checked it on all PRs when you suggested it on one of them :p
I'll include this as a check for the final doc revamp PR.

Comment (awaelchli) [reply docs/source/starter/style_guide.rst]: No worries. I'm just making a little fun of git 😄

File: docs/source/starter/style_guide.rst [hunk 11]
```diff
@@ -147,26 +153,33 @@ In practice, this code looks like:
...
-        x, y = ...
+    def forward(self, x):
+        embeddings = self.encoder(x)
+        return embeddings
+
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
         z = self(x)  # < ---------- instead of self.encoder(x)
         pred = self.decoder(z)
```
Comment (justusschock) [inline docs/source/starter/style_guide.rst:184]: a question in general (even though you did not change this): If we need to call `self(x)` instead of `self.encoder(x)`, can we still call `self.decoder(z)` or would this also have to go to `forward`? I don't see a difference there. If there is one, it probably has to be explained here...

Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:184]: yes.. this isn't required anymore. I think.
since batch is already splitter at this point, so calling self(), or self.encoder(x) won't be a problem of any sort.

for DP model is warpped around `model=DataParallel(LightningParallelModule(LightningModule)))`

`LightningParallelModule` takes care of which method to call, while `DataParallel` splits the batch and calls LightningParallelModule's forward which then calls `training_step` while training with the splitted batch.

cc: @akihironitta

Comment (rohitgr7) [reply docs/source/starter/style_guide.rst:184]: this is what I could find from my investigation. But if anyone can confirm this, I'll update this.

Comment (justusschock) [reply docs/source/starter/style_guide.rst:184]: I think so as well :)

Comment (rohitgr7) [reply docs/source/starter/style_guide.rst]: thanks @justusschock for bringing this up :)

File: docs/source/starter/style_guide.rst (hunk 7 above)
Comment (akihironitta) [reply docs/source/starter/style_guide.rst:43]: I agree with @ananthsub.

@rohitgr7 How about explicitly mentioning here that users are discouraged to define a "model" like the example (although users can still write that way)?

File: docs/source/starter/style_guide.rst [hunk 12]
```diff
@@ -49,7 +51,7 @@ Here's a LightningModule that defines a system:
 .. testcode::
 
     class LitModel(LightningModule):
-        def __init__(self, encoder: nn.Module = None, decoder: nn.Module = None):
+        def __init__(self, encoder, decoder=None):
```
Comment (akihironitta) [inline docs/source/starter/style_guide.rst:54]: nit
```suggestion
        def __init__(self, encoder, decoder):
```

**Review by akihironitta — APPROVED** [review]

File: docs/source/starter/style_guide.rst [hunk 13]
```diff
@@ -24,41 +26,60 @@ In Lightning we differentiate between a system and a model.
...
     class LitModel(LightningModule):
-        def __init__(self, num_layers: int = 3):
+        def __init__(self):
             super().__init__()
             self.layer_1 = nn.Linear()
             self.layer_2 = nn.Linear()
             self.layer_3 = nn.Linear()
 
-Here's a LightningModule that defines a system:
+Here's a LightningModule that defines a system. This is how we recommend.
```
Comment (carmocca) [inline docs/source/starter/style_guide.rst:49]:
```suggestion
Here's a LightningModule that defines a system. This structure is what we recommend as a best practice.
```

**Review by carmocca — APPROVED** [review]

File: docs/source/starter/style_guide.rst [hunk 14]
```diff
@@ -1,8 +1,9 @@
 ###########
-Style guide
+Style Guide
 ###########
-A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo,
-finding a lightning module and knowing exactly where to look to find the things you care about.
+
+A main goal of Lightning is to improve readability and reproducibility. Imagine looking into any GitHub repo or a research project,
+finding a lightning module, and knowing exactly where to look to find the things you care about.
```
Comment (justusschock) [inline docs/source/starter/style_guide.rst:6]:
```suggestion
finding a :class:`~pytorch_lightning.core.lightning.LightningModule`, and knowing exactly where to look to find the things you care about.
```

**Review by justusschock — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: docs/source/starter/style_guide.rst [hunk 15]
```diff
@@ -24,41 +26,60 @@ In Lightning we differentiate between a system and a model.
...
     class LitModel(LightningModule):
-        def __init__(self, num_layers: int = 3):
+        def __init__(self):
             super().__init__()
             self.layer_1 = nn.Linear()
             self.layer_2 = nn.Linear()
             self.layer_3 = nn.Linear()
 
-Here's a LightningModule that defines a system:
+Here's a LightningModule that defines a system. This structure is what we recommend as a best practice.
```
Comment (ananthsub) [inline docs/source/starter/style_guide.rst:49]: This should come first in the docs since it's the recommended path to follow 

Somewhere here we should emphasize that this distinction improves modularity, which has a slew of benefits (testability, reduced dependencies, easier to refactor)

File: docs/source/starter/style_guide.rst [hunk 16]
```diff
@@ -24,41 +26,62 @@ In Lightning we differentiate between a system and a model.
...
 A model is something like a resnet18, RNN, etc.
 
-A system defines how a collection of models interact with each other. Examples of this are:
+A system defines how a collection of models interact with each other with user-defined training/evaluation logic. Examples of this are:
 
 * GANs
 * Seq2Seq
 * BERT
-* etc
+* etc...
```
Comment (Borda) [inline docs/source/starter/style_guide.rst:34]:
```suggestion
* etc.
```
I guess it is "etc." or "..." but not both
